pkg/controller/common/util.go

```diff
@@ -1887,25 +1898,74 @@ func GetSnapshotClassForSmartClone(dvName string, targetPvcStorageClassName *str
 		return "", nil
 	}
 
-	// List the snapshot classes
-	scs := &snapshotv1.VolumeSnapshotClassList{}
-	if err := client.List(context.TODO(), scs); err != nil {
-		logger.Info("Cannot list snapshot classes, falling back to host assisted clone")
+	vscName, err := GetVolumeSnapshotClass(context.TODO(), client, pvc, targetStorageClass.Provisioner, snapshotClassName, logger, recorder)
+	if err != nil {
 		return "", err
 	}
-	for _, snapshotClass := range scs.Items {
-		// Validate association between snapshot class and storage class
-		if snapshotClass.Driver == targetStorageClass.Provisioner {
+	if vscName != nil {
+		if pvc != nil {
 			logger.Info("smart-clone is applicable for datavolume", "datavolume",
-				dvName, "snapshot class", snapshotClass.Name)
-			return snapshotClass.Name, nil
+				pvc.Name, "snapshot class", *vscName)
 		}
+		return *vscName, nil
 	}
 
 	logger.Info("Could not match snapshotter with storage class, falling back to host assisted clone")
 	return "", nil
 }
 
+// GetVolumeSnapshotClass looks up the snapshot class based on the driver and an optional specific name
+func GetVolumeSnapshotClass(ctx context.Context, c client.Client, pvc *corev1.PersistentVolumeClaim, driver string, snapshotClassName *string, log logr.Logger, recorder record.EventRecorder) (*string, error) {
+	logger := log.WithName("GetVolumeSnapshotClass").V(3)
+
+	logEvent := func(message, vscName string) {
+		logger.Info(message, "name", vscName)
+		if pvc != nil {
+			msg := fmt.Sprintf("%s %s", message, vscName)
+			recorder.Event(pvc, corev1.EventTypeNormal, VolumeSnapshotClassSelected, msg)
+		}
+	}
+
+	if snapshotClassName != nil {
+		vsc := &snapshotv1.VolumeSnapshotClass{}
+		if err := c.Get(context.TODO(), types.NamespacedName{Name: *snapshotClassName}, vsc); err != nil {
+			return nil, err
+		}
+		if vsc.Driver == driver {
+			logEvent(MessageStorageProfileVolumeSnapshotClassSelected, vsc.Name)
+			return &vsc.Name, nil
+		}
+		return nil, nil
+	}
+
+	vscList := &snapshotv1.VolumeSnapshotClassList{}
+	if err := c.List(ctx, vscList); err != nil {
+		if meta.IsNoMatchError(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	var candidates []string
+	for _, vsc := range vscList.Items {
+		if vsc.Driver == driver {
+			if vsc.Annotations[AnnDefaultSnapshotClass] == "true" {
+				logEvent(MessageDefaultVolumeSnapshotClassSelected, vsc.Name)
+				return &vsc.Name, nil
+			}
+			candidates = append(candidates, vsc.Name)
+		}
+	}
+
+	if len(candidates) > 0 {
+		sort.Strings(candidates)
```

Comment:
Not sure what's the best behavior here,  I don't think it's a huge deal but have to make sure we know this
k8s does this:
https://github.com/kubevirt/containerized-data-importer/blob/539280843c306e415ec072dd87a7181ed1e364cf/pkg/controller/common/util.go#L520-L529
but the external snapshotter behaves differently:
https://github.com/kubernetes-csi/external-snapshotter/blob/ff71329d8cc08dca53194f8d4568ed53e69ecb82/pkg/common-controller/snapshot_controller.go#L1362-L1381

Reply:
The k8s and external snapshotter code are for getting the default storage/snapshot class, but we deal here with what to return when there is no such default.